File: app/src/main/java/com/czbix/v2ex/model/Node.java
package com.czbix.v2ex.model;

import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.czbix.v2ex.common.exception.FatalException;
import com.czbix.v2ex.network.RequestHelper;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;

import java.text.Collator;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Node extends Page implements Comparable<Node> {
    private static final Pattern PATTERN = Pattern.compile("/go/(.+?)(?:\\W|$)");
    private static final Collator COLLATOR = Collator.getInstance(Locale.CHINA);

    private final int mId;
    private final String mTitle;
    private final String mName;
    private final String mTitleAlternative;
    private final int mTopics;
    private final Avatar mAvatar;
    private final boolean mHasInfo;

    public Node(String title, int id, Avatar avatar, String name, String alternative,
                int topics) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(name));
        mId = id;

        mTitle = title;
        mAvatar = avatar;
        mName = name;
        mTitleAlternative = alternative;
        mTopics = topics;

        mHasInfo = !Strings.isNullOrEmpty(title);
    }

    public int getId() {
        return mId;
    }

    @Override
    public String getTitle() {
        return mTitle;
    }

    @Nullable
    public Avatar getAvatar() {
        return mAvatar;
    }

    public String getName() {
        return mName;
    }

    public String getTitleAlternative() {
        return mTitleAlternative;
    }

    public int getTopics() {
        return mTopics;
    }

    public boolean hasInfo() {
        return mHasInfo;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Node)) return false;
        Node node = (Node) o;
        return Objects.equal(mId, node.mId) &&
                Objects.equal(mTopics, node.mTopics) &&
                Objects.equal(mName, node.mName) &&
                Objects.equal(mTitleAlternative, node.mTitleAlternative) &&
                Objects.equal(mAvatar, node.mAvatar);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(mId, mName, mTitleAlternative, mTopics, mAvatar);
    }

    public static String buildUrlByName(String name) {
        return RequestHelper.BASE_URL + "/go/" + name;
    }

    public static String getNameFromUrl(String url) {
        final Matcher matcher = PATTERN.matcher(url);
        if (!matcher.find()) {
            throw new FatalException("match name for node failed: " + url);
        }
        return matcher.group(1);
    }

    @Override
    public String getUrl() {
        return buildUrlByName(getName());
    }

    @Override
    public int compareTo(@NonNull Node another) {
        return COLLATOR.compare(getTitle(), another.getTitle());
    }

    public static class Builder {
        private int mId;
        private String mTitle;
        private Avatar mAvatar;
        private String mName;
        private String mTitleAlternative;
        private int mTopics;

        public Builder setId(int id) {
            mId = id;
            return this;
        }

        public Builder setTitle(String title) {
            mTitle = title;
            return this;
        }

        public Builder setAvatar(Avatar avatar) {
            mAvatar = avatar;
            return this;
        }

        public Builder setName(String name) {
            mName = name;
            return this;
        }

        public Builder setTitleAlternative(String titleAlternative) {
            mTitleAlternative = titleAlternative;
            return this;
        }

        public Builder setTopics(int topics) {
            mTopics = topics;
            return this;
        }

        public Node createNode() {
            return new Node(mTitle, mId, mAvatar, mName, mTitleAlternative, mTopics);
        }
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.mId);
        dest.writeString(this.mTitle);
        dest.writeString(this.mName);
        dest.writeString(this.mTitleAlternative);
        dest.writeInt(this.mTopics);
        dest.writeParcelable(this.mAvatar, 0);
        dest.writeByte(mHasInfo ? (byte) 1 : (byte) 0);
    }

    protected Node(Parcel in) {
        this.mId = in.readInt();
        this.mTitle = in.readString();
        this.mName = in.readString();
        this.mTitleAlternative = in.readString();
        this.mTopics = in.readInt();
        this.mAvatar = in.readParcelable(Avatar.class.getClassLoader());
        this.mHasInfo = in.readByte() != 0;
    }

    public static final Creator<Node> CREATOR = new Creator<Node>() {
        public Node createFromParcel(Parcel source) {
            return new Node(source);
        }

        public Node[] newArray(int size) {
            return new Node[size];
        }
    };
}

File: app/src/main/kotlin/com/czbix/v2ex/model/Node.kt
package com.czbix.v2ex.model

import android.os.Parcel
import android.os.Parcelable
import com.czbix.v2ex.common.exception.FatalException
import com.czbix.v2ex.network.RequestHelper
import com.czbix.v2ex.ui.widget.ExArrayAdapter
import com.google.common.base.Objects
import com.google.common.base.Preconditions
import com.google.common.base.Strings
import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import java.text.Collator
import java.util.*
import java.util.concurrent.ExecutionException
import java.util.regex.Pattern

class Node : Page, Comparable<Node>, ExArrayAdapter.Filterable, Favable {
    val id: Int
    private val mTitle: String?
    val name: String
    val titleAlternative: String?
    val topics: Int
    val avatar: Avatar?
    private val mHasInfo: Boolean

    constructor(title: String?, id: Int, avatar: Avatar?, name: String, alternative: String?,
                topics: Int) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(name))
        this.id = id

        mTitle = title
        this.avatar = avatar
        this.name = name
        titleAlternative = alternative
        this.topics = topics

        mHasInfo = !Strings.isNullOrEmpty(title)
    }

    override fun getTitle(): String? {
        return mTitle
    }

    fun hasInfo(): Boolean {
        return mHasInfo
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Node) return false
        return Objects.equal(id, other.id) &&
                Objects.equal(topics, other.topics) &&
                Objects.equal(name, other.name) &&
                Objects.equal(titleAlternative, other.titleAlternative) &&
                Objects.equal(avatar, other.avatar)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id, name, titleAlternative, topics, avatar)
    }

    override fun getUrl(): String {
        return buildUrlByName(name)
    }

    override fun filter(query: String): Boolean {
        if (Strings.isNullOrEmpty(query)) {
            return true
        }
        if (name.contains(query) || (mTitle != null && mTitle.contains(query))) {
            return true
        }
        if (titleAlternative != null && titleAlternative.contains(query)) {
            return true
        }
        return false
    }

    override fun compareTo(other: Node): Int {
        return COLLATOR.compare(title, other.title)
    }

    override fun getFavUrl(token: String): String {
        return String.format("%s/favorite/node/%d?once=%s", RequestHelper.BASE_URL, id, token)
    }

    override fun getUnFavUrl(token: String): String {
        return String.format("%s/unfavorite/node/%d?once=%s", RequestHelper.BASE_URL, id, token)
    }

    class Builder {
        private var mId: Int = 0
        private var mTitle: String? = null
        private var mAvatar: Avatar? = null
        private var mName: String? = null
        private var mTitleAlternative: String? = null
        private var mTopics: Int = 0

        fun setId(id: Int): Builder {
            mId = id
            return this
        }

        fun setTitle(title: String): Builder {
            mTitle = title
            return this
        }

        fun setAvatar(avatar: Avatar): Builder {
            mAvatar = avatar
            return this
        }

        fun setName(name: String): Builder {
            mName = name
            return this
        }

        fun setTitleAlternative(titleAlternative: String): Builder {
            mTitleAlternative = titleAlternative
            return this
        }

        fun setTopics(topics: Int): Builder {
            mTopics = topics
            return this
        }

        fun createNode(): Node {
            try {
                val name = checkNotNull(mName)
                if (mTitle == null) {
                    return Node(null, mId, mAvatar, name, mTitleAlternative, mTopics)
                }
                return CACHE[name, { Node(mTitle, mId, mAvatar, name, mTitleAlternative, mTopics) }]
            } catch (e: ExecutionException) {
                throw FatalException(e)
            }

        }

        companion object {
            private val CACHE: Cache<String, Node>

            init {
                CACHE = CacheBuilder.newBuilder()
                        .softValues()
                        .initialCapacity(32)
                        .maximumSize(128)
                        .build<String, Node>()
            }
        }
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(this.id)
        dest.writeString(this.mTitle)
        dest.writeString(this.name)
        dest.writeString(this.titleAlternative)
        dest.writeInt(this.topics)
        dest.writeParcelable(this.avatar, 0)
        dest.writeByte(if (mHasInfo) 1.toByte() else 0.toByte())
    }

    private constructor(`in`: Parcel) {
        this.id = `in`.readInt()
        this.mTitle = `in`.readString()
        this.name = `in`.readString()
        this.titleAlternative = `in`.readString()
        this.topics = `in`.readInt()
        this.avatar = `in`.readParcelable<Avatar>(Avatar::class.java.classLoader)
        this.mHasInfo = `in`.readByte().toInt() != 0
    }

    companion object {
        private val PATTERN = Pattern.compile("/go/(.+?)(?:\\W|$)")
        private val COLLATOR = Collator.getInstance(Locale.CHINA)

        fun buildUrlByName(name: String): String {
            return RequestHelper.BASE_URL + "/go/" + name
        }

        fun getNameFromUrl(url: String): String {
            val matcher = PATTERN.matcher(url)
            Preconditions.checkState(matcher.find(), "match name for node failed: " + url)
            return matcher.group(1)
        }

        val CREATOR: Parcelable.Creator<Node> = object : Parcelable.Creator<Node> {
            override fun createFromParcel(source: Parcel): Node {
                return Node(source)
            }

            override fun newArray(size: Int): Array<Node?> {
                return arrayOfNulls(size)
            }
        }
    }
}
